docs/man/memcached_auto.3
.TH "MEMCACHED_AUTO" "3" "September 16, 2012" "1.0.10" "libmemcached"
.SH NAME
memcached_auto \- Incrementing and Decrementing Values
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.SH SYNOPSIS
.sp
#include <libmemcached/memcached.h>
.INDENT 0.0
.TP
.B memcached_return_t memcached_increment(memcached_st\fI\ *ptr\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, uint32_t\fI\ offset\fP, uint64_t\fI\ *value\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_decrement(memcached_st\fI\ *ptr\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, uint32_t\fI\ offset\fP, uint64_t\fI\ *value\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_increment_with_initial(memcached_st\fI\ *ptr\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, uint64_t\fI\ offset\fP, uint64_t\fI\ initial\fP, time_t\fI\ expiration\fP, uint64_t\fI\ *value\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_decrement_with_initial(memcached_st\fI\ *ptr\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, uint64_t\fI\ offset\fP, uint64_t\fI\ initial\fP, time_t\fI\ expiration\fP, uint64_t\fI\ *value\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_increment_by_key(memcached_st\fI\ *ptr\fP, const char\fI\ *group_key\fP, size_t\fI\ group_key_length\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, uint32_t\fI\ offset\fP, uint64_t\fI\ *value\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_decrement_by_key(memcached_st\fI\ *ptr\fP, const char\fI\ *group_key\fP, size_t\fI\ group_key_length\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, uint32_t\fI\ offset\fP, uint64_t\fI\ *value\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_increment_with_initial_by_key(memcached_st\fI\ *ptr\fP, const char\fI\ *group_key\fP, size_t\fI\ group_key_length\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, uint64_t\fI\ offset\fP, uint64_t\fI\ initial\fP, time_t\fI\ expiration\fP, uint64_t\fI\ *value\fP)
.UNINDENT
.INDENT 0.0
.TP
.B memcached_return_t memcached_decrement_with_initial_by_key(memcached_st\fI\ *ptr\fP, const char\fI\ *group_key\fP, size_t\fI\ group_key_length\fP, const char\fI\ *key\fP, size_t\fI\ key_length\fP, uint64_t\fI\ offset\fP, uint64_t\fI\ initial\fP, time_t\fI\ expiration\fP, uint64_t\fI\ *value\fP)
.UNINDENT
.sp
Compile and link with \-lmemcached
.SH DESCRIPTION
.sp
\fImemcached(1)\fP servers have the ability to increment and decrement keys
(overflow and underflow are not detected). This gives you the ability to use
memcached to generate shared sequences of values.
.sp
memcached_increment takes a key and keylength and increments the value by
the offset passed to it. The value is then returned via the uint32_t
value pointer you pass to it.
.sp
memcached_decrement takes a key and keylength and decrements the value by
the offset passed to it. The value is then returned via the uint32_t
value pointer you pass to it.
.sp
memcached_increment_with_initial takes a key and keylength and increments
the value by the offset passed to it. If the object specified by key does
not exist, one of two things may happen: If the expiration value is
MEMCACHED_EXPIRATION_NOT_ADD, the operation will fail. For all other
expiration values, the operation will succeed by seeding the value for that
key with a initial value to expire with the provided expiration time. The
flags will be set to zero.The value is then returned via the uint32_t
value pointer you pass to it.
.sp
memcached_decrement_with_initial takes a key and keylength and decrements
the value by the offset passed to it. If the object specified by key does
not exist, one of two things may happen: If the expiration value is
MEMCACHED_EXPIRATION_NOT_ADD, the operation will fail. For all other
expiration values, the operation will succeed by seeding the value for that
key with a initial value to expire with the provided expiration time. The
flags will be set to zero.The value is then returned via the uint32_t
value pointer you pass to it.
.sp
\fI\%memcached_increment_by_key()\fP, \fI\%memcached_decrement_by_key()\fP,
\fI\%memcached_increment_with_initial_by_key()\fP, and
\fI\%memcached_decrement_with_initial_by_key()\fP are master key equivalents of the above.
.SH RETURN
.sp
A value of type \fBmemcached_return_t\fP  is returned.
On success that value will be \fBMEMCACHED_SUCCESS\fP.
Use memcached_strerror to translate this value to a printable string.
.SH HOME
.sp
To find out more information please check:
\fI\%http://libmemcached.org/\fP
.SH SEE ALSO
.sp
\fImemcached(1)\fP \fIlibmemcached(3)\fP \fImemcached_strerror(3)\fP
.SH AUTHOR
Brian Aker
.SH COPYRIGHT
2011-2012, Brian Aker DataDifferential, http://datadifferential.com/
.\" Generated by docutils manpage writer.
.\" 
.


ChangeLog
1.0.11 Sun Sep 16 20:32:13 EDT 2012
* Removed custom version of memcached.
* Updated hardening rules.
* Fixed a case where the return error from a socket connection differred from that of a TCP/IP socket.

1.0.10 Sun Jul 29 21:50:15 PDT 2012
* --disable-assert has been removed from configure, and --enable-assert has been added in its place.
* Compiling fixes for Clang on OSX Mountain Lion.

1.0.9 Wed Jul  4 22:46:19 EDT 2012
* Faster close on socket.
* Instance allocation is now seperated from server interface. This should allow for a better preservation of ABI compliance from now on.
* Fix close on exec bug.
* Numerous other bug fixes.

1.0.8 Tue May 22 15:06:04 EDT 2012
* Added support for setting options via ENV variable LIBMEMCACHED
* Fix corner case on last used result.

1.0.7 Sat Apr 28 00:48:29 PDT 2012
* Add API call for exist calls.
* Update all license files to be BSD.

1.0.6 Sat Apr  7 18:26:49 PDT 2012
* Fixes for gcc 4.7, lp:961812
* Fix for restart issue that happens under testing.
* Fix for lp:962815.
* Support for transparent AES encryption.

1.0.5 Tue Mar 13 22:56:47 PDT 2012
* Fixes for OSX.
* Version is now parsed directly in the parser, which makes buffered operations now work with it..
* memstat has been extended so that it can be used to find the version of the server.
* Update documentation.
* Fixes for compile issues on Debian and Ubuntu

1.0.4 Thu Jan 26 22:33:54 PST 2012
* Fix for memcached_dump().
* Additional testing for memcached_stat_execute().

1.0.3 Sun Jan  8 18:11:36 PST 2012
* Increased size of sort buffer used during Ketama.
* Added support for new behavior to handle dead servers.
* Overall haul of UDP IO.
* Fixed C compile issue with memcached_exist()
* Numerous bug fixes.
* Clang support for OSX.
* All commands now using vector send support.

1.0.2 Mon Oct 24 08:00:48 PDT 2011
* Dropped libmemcached/memcached_util.h (undocumented header file)
* Added memcached_touch() and memcached_touch_by_key()
* UDP support restructured to toggle on a complete memcached_st structure.

0.53 Mon Sep 26 20:50:33 PDT 2011
  * Fix for FreeBSD/OpenBSD and -lm
  * Added memcached_exist()
  * Fix for memory when using config test.
  * CLI gained --quiet

0.52 Sun Sep 11 22:16:08 PDT 2011
  * Build fixes for Ubuntu/Suse.
  * Fixes for OSX Lion.
  * Bug fix for looping back through dns lookups under certain failures.
  * Fixes related to dead server failures.

0.51 Thu Jul 21 11:08:58 PDT 2011
  * memcached_callback_set() now takes its data argument as const
  * Update to tests.
  * Fix in parser for port number.

0.50 Mon Jun 20 10:36:57 PDT 2011
 * Updates to C++ interface 
 * Custom free allocators need to now check for value before calling free.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).
 * Fix for stats structure.
 * Updates to documentation.
 * memcached_fetch_result() now uses the internal result when available (about 25 to 50% faster).

0.49 Thu Apr 14 08:43:37 PDT 2011
  * Fix calls to auto methods so that if value is not passed in nothing bad happens.
  * New parser calls for generating memcached_st objects.
  * New error system.
  * New flow control for messages means faster get/set calls.
  * Added new documentation system.
  * A behavior change has been now made that if you specify a weight for any server, we enable the weight flag and do weight balancing.  
  * Added MEMCACHED_BEHAVIOR_REMOVE_FAILED_SERVERS to simplify the setting of AUTO REJECT for servers.

0.48 Tue Mar 15 23:05:18 PDT 2011
  * Fix memory leak in server parse.
  * Move test framework out to be its own library (easier to work with Gearman).


0.47 Wed Feb 23 18:46:48 PST 2011
  * Additional fixes for OpenBSD.
  * Bug fix 677609, 456080.
  * SIGPIPE fix for Linux send(). 
  * memcapable can now test ascii or binary based on flags.
  * Additional build fixes for SASL.


0.46 Mon Feb 14 10:28:01 PST 2011
  * Fixes a number of corner case bugs.
  * Fixes related to OpenBSD.
  * Better testing for protocol version.
  * Removes special case infinite wait on blocking setup.

0.45 Tue Feb  8 16:02:06 PST 2011
  * Add support for systemtap

0.44 Wed Sep 22 21:57:57 PDT 2010
  * Windows bug fixes.
  * Hudson port support in test harness.
  * Improved portability of test hanrness.
  * SASL fixes.

0.43 Wed Jul 28 16:29:47 PDT 2010
  * Added --args to memstat so that a greater range of values can be returned.
  * Prelimanary support for Windows.
  * memcached_stat_execute() merged. 

0.42 Tue Jul  6 12:29:50 PDT 2010
  * Mistake in libtool caused issue with library version

0.41 Wed Jun 30 13:15:19 PDT 2010
  * Added --file for memcat.
  * Added limemcached_ping() to libmemcached_util
  * Bugfix for some cases where connect would have issues with timeout.
  * Wrong value for errno given as error on an IO failure inside of poll.
  * Bug fix for issue where multiple interfaces with bad DNS were not being caught.

0.40 Thu Apr 22 19:01:25 PDT 2010
  * Placed retry logic in for busted resolvers
  * Add an ignore for SIGPIPE to solve OSX issues.
  * A couple of fixed for memcached_light server.
  * Updated to debug mode to track io_wait

0.39 Tue Apr  6 12:35:13 PDT 2010
  * Add support for prefix keys to binary protocol.
  * Remove the undocumented call memcached_server_remove().
  * The undocumented call memcached_server_by_key() now returns const.
  * memcached_server_error_reset() has been deprecated.
  * memcached_server_list() has been deprecated. Use memcached_server_cursor() to walk the servers found in a memcached_st() structure.
  * memcached_verbosity() can now be run concurrently with other operations.
  * SASL support.
  * Fixes memory leak found in EJECT HOSTS.

0.38 Wed Feb 10 09:40:33 PST 2010
  * C++ interface for libhashkit.
  * Modified memcached_set_memory_allocators() so that it requires a context pointer.
  * memcached_clone() now runs 5 times faster.
  * Functions used for callbacks are now given const memcached_st.

libmemcached/memcached_auto.c
/* LibMemcached
 * Copyright (C) 2006-2009 Brian Aker
 * All rights reserved.
 *
 * Use and distribution licensed under the BSD license.  See
 * the COPYING file in the parent directory for full text.
 *
 * Summary: Methods for adding or decrementing values from an object in memcached
 *
 */

#include "common.h"

static memcached_return_t memcached_auto(memcached_st *ptr,
                                         const char *verb,
                                         const char *master_key, size_t master_key_length,
                                         const char *key, size_t key_length,
                                         uint64_t offset,
                                         uint64_t *value)
{
  size_t send_length;
  memcached_return_t rc;
  char buffer[MEMCACHED_DEFAULT_COMMAND_SIZE];
  unsigned int server_key;
  bool no_reply= ptr->flags.no_reply;

  unlikely (ptr->hosts == NULL || ptr->number_of_hosts == 0)
    return MEMCACHED_NO_SERVERS;

  if (ptr->flags.verify_key && (memcached_key_test((const char **)&key, &key_length, 1) == MEMCACHED_BAD_KEY_PROVIDED))
    return MEMCACHED_BAD_KEY_PROVIDED;

  server_key= memcached_generate_hash(ptr, master_key, master_key_length);

  send_length= (size_t)snprintf(buffer, MEMCACHED_DEFAULT_COMMAND_SIZE,
                                "%s %s%.*s %" PRIu64 "%s\r\n", verb,
                                ptr->prefix_key,
                                (int)key_length, key,
                                offset, no_reply ? " noreply" : "");
  unlikely (send_length >= MEMCACHED_DEFAULT_COMMAND_SIZE)
    return MEMCACHED_WRITE_FAILURE;

  rc= memcached_do(&ptr->hosts[server_key], buffer, send_length, 1);
  if (no_reply || rc != MEMCACHED_SUCCESS)
    return rc;

  rc= memcached_response(&ptr->hosts[server_key], buffer, MEMCACHED_DEFAULT_COMMAND_SIZE, NULL);

  /*
    So why recheck responce? Because the protocol is brain dead :)
    The number returned might end up equaling one of the string
    values. Less chance of a mistake with strncmp() so we will
    use it. We still called memcached_response() though since it
    worked its magic for non-blocking IO.
  */
  if (!strncmp(buffer, "ERROR\r\n", 7))
  {
    *value= 0;
    rc= MEMCACHED_PROTOCOL_ERROR;
  }
  else if (!strncmp(buffer, "NOT_FOUND\r\n", 11))
  {
    *value= 0;
    rc= MEMCACHED_NOTFOUND;
  }
  else
  {
    *value= strtoull(buffer, (char **)NULL, 10);
    rc= MEMCACHED_SUCCESS;
  }

  return rc;
}

static memcached_return_t binary_incr_decr(memcached_st *ptr, uint8_t cmd,
                                           const char *master_key, size_t master_key_length,
                                           const char *key, size_t key_length,
                                           uint64_t offset, uint64_t initial,
                                           uint32_t expiration,
                                           uint64_t *value)
{
  unsigned int server_key;
  bool no_reply= ptr->flags.no_reply;

  unlikely (ptr->hosts == NULL || ptr->number_of_hosts == 0)
    return MEMCACHED_NO_SERVERS;

  server_key= memcached_generate_hash(ptr, master_key, master_key_length);

  if (no_reply)
  {
    if(cmd == PROTOCOL_BINARY_CMD_DECREMENT)
      cmd= PROTOCOL_BINARY_CMD_DECREMENTQ;
    if(cmd == PROTOCOL_BINARY_CMD_INCREMENT)
      cmd= PROTOCOL_BINARY_CMD_INCREMENTQ;
  }
  protocol_binary_request_incr request= {.bytes= {0}};

  request.message.header.request.magic= PROTOCOL_BINARY_REQ;
  request.message.header.request.opcode= cmd;
  request.message.header.request.keylen= htons((uint16_t) key_length);
  request.message.header.request.extlen= 20;
  request.message.header.request.datatype= PROTOCOL_BINARY_RAW_BYTES;
  request.message.header.request.bodylen= htonl((uint32_t) (key_length + request.message.header.request.extlen));
  request.message.body.delta= htonll(offset);
  request.message.body.initial= htonll(initial);
  request.message.body.expiration= htonl((uint32_t) expiration);

  if ((memcached_do(&ptr->hosts[server_key], request.bytes,
                    sizeof(request.bytes), 0)!=MEMCACHED_SUCCESS) ||
      (memcached_io_write(&ptr->hosts[server_key], key, key_length, 1) == -1))
  {
    memcached_io_reset(&ptr->hosts[server_key]);
    return MEMCACHED_WRITE_FAILURE;
  }

  if (no_reply)
    return MEMCACHED_SUCCESS;
  return memcached_response(&ptr->hosts[server_key], (char*)value, sizeof(*value), NULL);
}

memcached_return_t memcached_increment(memcached_st *ptr,
                                       const char *key, size_t key_length,
                                       uint32_t offset,
                                       uint64_t *value)
{
  return memcached_increment_by_key(ptr, key, key_length, key, key_length, offset, value);
}

memcached_return_t memcached_decrement(memcached_st *ptr,
                                       const char *key, size_t key_length,
                                       uint32_t offset,
                                       uint64_t *value)
{
  return memcached_decrement_by_key(ptr, key, key_length, key, key_length, offset, value);
}

memcached_return_t memcached_increment_by_key(memcached_st *ptr,
                                              const char *master_key, size_t master_key_length,
                                              const char *key, size_t key_length,
                                              uint64_t offset,
                                              uint64_t *value)
{
  memcached_return_t rc= memcached_validate_key_length(key_length, ptr->flags.binary_protocol);
  unlikely (rc != MEMCACHED_SUCCESS)
    return rc;

  LIBMEMCACHED_MEMCACHED_INCREMENT_START();
  if (ptr->flags.binary_protocol)
  {
    rc= binary_incr_decr(ptr, PROTOCOL_BINARY_CMD_INCREMENT,
                         master_key, master_key_length, key, key_length,
                         (uint64_t)offset, 0, MEMCACHED_EXPIRATION_NOT_ADD,
                         value);
  }
  else
  {
     rc= memcached_auto(ptr, "incr", master_key, master_key_length, key, key_length, offset, value);
  }

  LIBMEMCACHED_MEMCACHED_INCREMENT_END();

  return rc;
}

memcached_return_t memcached_decrement_by_key(memcached_st *ptr,
                                              const char *master_key, size_t master_key_length,
                                              const char *key, size_t key_length,
                                              uint64_t offset,
                                              uint64_t *value)
{
  memcached_return_t rc= memcached_validate_key_length(key_length, ptr->flags.binary_protocol);
  unlikely (rc != MEMCACHED_SUCCESS)
    return rc;

  LIBMEMCACHED_MEMCACHED_DECREMENT_START();
  if (ptr->flags.binary_protocol)
    rc= binary_incr_decr(ptr, PROTOCOL_BINARY_CMD_DECREMENT,
                         master_key, master_key_length, key, key_length,
                         (uint64_t)offset, 0, MEMCACHED_EXPIRATION_NOT_ADD,
                         value);
  else
    rc= memcached_auto(ptr, "decr", master_key, master_key_length, key, key_length, offset, value);

  LIBMEMCACHED_MEMCACHED_DECREMENT_END();

  return rc;
}

memcached_return_t memcached_increment_with_initial(memcached_st *ptr,
                                                    const char *key,
                                                    size_t key_length,
                                                    uint64_t offset,
                                                    uint64_t initial,
                                                    time_t expiration,
                                                    uint64_t *value)
{
  return memcached_increment_with_initial_by_key(ptr, key, key_length,
                                                 key, key_length,
                                                 offset, initial, expiration, value);
}

memcached_return_t memcached_increment_with_initial_by_key(memcached_st *ptr,
                                                         const char *master_key,
                                                         size_t master_key_length,
                                                         const char *key,
                                                         size_t key_length,
                                                         uint64_t offset,
                                                         uint64_t initial,
                                                         time_t expiration,
                                                         uint64_t *value)
{
  memcached_return_t rc= memcached_validate_key_length(key_length, ptr->flags.binary_protocol);
  unlikely (rc != MEMCACHED_SUCCESS)
    return rc;

  LIBMEMCACHED_MEMCACHED_INCREMENT_WITH_INITIAL_START();
  if (ptr->flags.binary_protocol)
    rc= binary_incr_decr(ptr, PROTOCOL_BINARY_CMD_INCREMENT,
                         master_key, master_key_length, key, key_length,
                         offset, initial, (uint32_t)expiration,
                         value);
  else
    rc= MEMCACHED_PROTOCOL_ERROR;

  LIBMEMCACHED_MEMCACHED_INCREMENT_WITH_INITIAL_END();

  return rc;
}

memcached_return_t memcached_decrement_with_initial(memcached_st *ptr,
                                                    const char *key,
                                                    size_t key_length,
                                                    uint64_t offset,
                                                    uint64_t initial,
                                                    time_t expiration,
                                                    uint64_t *value)
{
  return memcached_decrement_with_initial_by_key(ptr, key, key_length,
                                                 key, key_length,
                                                 offset, initial, expiration, value);
}

memcached_return_t memcached_decrement_with_initial_by_key(memcached_st *ptr,
                                                           const char *master_key,
                                                           size_t master_key_length,
                                                           const char *key,
                                                           size_t key_length,
                                                           uint64_t offset,
                                                           uint64_t initial,
                                                           time_t expiration,
                                                           uint64_t *value)
{
  memcached_return_t rc= memcached_validate_key_length(key_length, ptr->flags.binary_protocol);
  unlikely (rc != MEMCACHED_SUCCESS)
    return rc;

  LIBMEMCACHED_MEMCACHED_INCREMENT_WITH_INITIAL_START();
  if (ptr->flags.binary_protocol)
  {
    rc= binary_incr_decr(ptr, PROTOCOL_BINARY_CMD_DECREMENT,
                         master_key, master_key_length, key, key_length,
                         offset, initial, (uint32_t)expiration,
                         value);
  }
  else
  {
    rc= MEMCACHED_PROTOCOL_ERROR;
  }

  LIBMEMCACHED_MEMCACHED_INCREMENT_WITH_INITIAL_END();

  return rc;
}

